File: 695/maxAreaOfIsland.scala
object Solution {
        
    def checkMax(grid: Array[Array[Int]], b: Array[Array[Boolean]], pos: (Int, Int)): Int = {
      if(pos._1 < 0 || pos._2 < 0 || pos._1 >= grid(pos._2).length || pos._2 >= grid.length) 0
      else if (grid(pos._2)(pos._1) == 0 || b(pos._2)(pos._1)) {
          b(pos._2)(pos._1) = true
          0
      } else {
          val (res1, res2, res3, res4) = (
              checkMax(grid, b, (pos._1 - 1, pos._2)), 
              checkMax(grid, b, (pos._1, pos._2 - 1)),
              checkMax(grid, b, (pos._1 + 1, pos._2)),
              checkMax(grid, b, (pos._1, pos._2 + 1))
          )
                       
          1 +  res1 + res2 + res3 + res4
      }
    }
    
    def maxAreaOfIsland(grid: Array[Array[Int]]): Int = {
        grid.headOption match {
          case Some(h) =>
            val height: Int              = grid.length
            val width:  Int              = h.length
            val b: Array[Array[Boolean]] = Array.ofDim[Boolean](height, width)
            var maxV: Int                = 0
            for(i <- 0 until height){
                for(j <- 0 until width){
                    maxV = checkMax(grid, b, (j, i)) match {
                        case x if x > maxV => x
                        case _ => maxV
                    }
                }
            }
            maxV
          case None    => 0   
        }
    } 
}

File: 695/maxAreaOfIsland.go
type Pos2d struct {
    x int
    y int
}

func checkMax(grid [][]int, b *[][] bool, pos Pos2d) int{

    if pos.x < 0 || pos.y < 0 || pos.x >= len(grid[pos.y]) || pos.y >= len(grid) {
      return 0
    } else if grid[pos.y][pos.x] == 0 || (*b)[pos.y][pos.x] {
      (*b)[pos.y][pos.x] = true
      return 0
    } else {
      res1 := 0
      res2 := 0
      res3 := 0
      res4 := 0

      (*b)[pos.y][pos.x] = true

      if pos.x > 0  && !(*b)[pos.y][pos.x - 1] {
        res1 = checkMax(grid, b, Pos2d{x: pos.x - 1, y: pos.y})
      }

      if pos.y > 0 && !(*b)[pos.y - 1][pos.x] {
        res2 = checkMax(grid, b, Pos2d{x: pos.x, y: pos.y - 1})
      }

      if pos.x < len(grid[pos.y]) - 1 && !(*b)[pos.y][pos.x + 1] {
        res3 = checkMax(grid, b, Pos2d{x: pos.x + 1, y: pos.y})
      }

      if pos.y < len(grid) - 1&& !(*b)[pos.y + 1][pos.x] {
        res4 = checkMax(grid, b, Pos2d{x: pos.x, y: pos.y + 1})
      }

      return (1 + res1 + res2 + res3 + res4)
    }
}

func maxAreaOfIsland(grid [][]int) int {
    b := [][]bool{}
    maxV := 0

    for i, h:= range grid {
        b = append(b, []bool{})
        for _, _ = range h {
            b[i] = append(b[i], false)
        }
    }

    for i, h := range grid {
        for j, _ := range h {
            va := checkMax(grid, &b, Pos2d{x: j, y: i})
            if va > maxV {
              maxV = va
            }
        }
    }

    return maxV
}
